src/main/java/com/microsoft/azure/relay/ClientWebSocket.java

```diff
@@ -173,36 +174,41 @@ boolean isOpen() {
 	 * @throws TimeoutException thrown when a complete message frame is not received within the timeout.
 	 */
 	CompletableFuture<ByteBuffer> readBinaryAsync(Duration timeout) {
-		
-		return CompletableFutureUtil.timedSupplyAsync(timeout, () -> {
-				LinkedList<byte[]> fragments = new LinkedList<byte[]>();
-				boolean receivedWholeMsg;
-				int messageSize = 0;
-				
-				do {
-					MessageFragment fragment = fragmentQueue.dequeueAsync().join();
-					if (fragment == null) {
-						break;
-					}				
-					messageSize += fragment.getBytes().length;
-					fragments.add(fragment.getBytes());
-					receivedWholeMsg = fragment.isEnd();
-				} 
-				while (!receivedWholeMsg);
-	
-				byte[] message = new byte[messageSize];
-				int offset = 0;
-				for (byte[] bytes : fragments) {
-					System.arraycopy(bytes, 0, message, offset, bytes.length);
-					offset += bytes.length;
+		LinkedList<byte[]> fragments = new LinkedList<byte[]>();
+		AtomicInteger messageSize = new AtomicInteger(0);
+		return readFragmentsAsync(fragments, messageSize, timeout).thenApply((nullResult) -> {
+			byte[] message = new byte[messageSize.get()];
```

Comment:
>get [](start = 41, length = 3)

I was going to say if you have readFragmentsAsync return a CompletableFuture<Integer> you might be able to remove the AtomicInteger here. You're the expert on async java now. I suspect it might make the code harder with the quasi-recursive magic that's being used here. However, returning an Integer would mean having to possibly allocate multiple Integers vs one AtomicInteger. Anyhow, just think about how it might look with CF<Int> instead and probably stick with what you have.